plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteVectorDistanceFunction.java

```diff
@@ -0,0 +1,159 @@
+                        Types.DOUBLE,
+                        Optional.of("double"),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty())));
+    }
+
+    private static Optional<ParameterizedExpression> rewrite(ConnectorExpression expression, RewriteContext<ParameterizedExpression> context)
+    {
+        if (expression instanceof Constant constant) {
+            Object value = constant.getValue();
+            LongArrayBlock block = (LongArrayBlock) value;
+            StringBuilder vector = new StringBuilder("'[");
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                if (block.isNull(i)) {
+                    return Optional.empty();
+                }
+                double doubleValue = longBitsToDouble(block.getLong(i));
+                if (!isSupportedVector(doubleValue)) {
+                    return Optional.empty();
+                }
+                vector.append(doubleValue);
+                if (i != block.getPositionCount() - 1) {
+                    vector.append(",");
+                }
+            }
+            vector.append("]'");
+            return Optional.of(new ParameterizedExpression(vector.toString(), ImmutableList.of()));
+        }
+        if (expression instanceof Call call && call.getFunctionName().equals(CAST_FUNCTION_NAME)) {
+            ConnectorExpression argument = getOnlyElement(call.getArguments());
+            if (argument instanceof Variable variable) {
+                JdbcTypeHandle typeHandle = ((JdbcColumnHandle) context.getAssignment(variable.getName())).getJdbcTypeHandle();
+                // TODO type.equals("vector") should be improved to support pushdown on vector type which is installed in other schemas
+                if (!typeHandle.jdbcTypeName().map(type -> type.equals("vector")).orElse(false)) {
+                    return Optional.empty();
+                }
+                return Optional.of(new ParameterizedExpression(variable.getName(), ImmutableList.of()));
+            }
+            return Optional.empty();
+        }
+        Optional<ParameterizedExpression> translatedRightArgument = context.rewriteExpression(expression);
+        if (translatedRightArgument.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(translatedRightArgument.orElseThrow());
+    }
+
+    private static boolean isArrayTypeWithRealOrDouble(Type type)
+    {
+        return type instanceof ArrayType arrayType && (arrayType.getElementType() == REAL || arrayType.getElementType() == DOUBLE);
+    }
+
+    private static boolean isSupportedVector(double value)
+    {
+        return !Double.isNaN(value) &&
+                !Double.isInfinite(value) &&
+                // pgvector throws an exception if the value is outside of float type
+                (value >= Float.MIN_VALUE && value <= Float.MAX_VALUE);
```

Comment:
For my understanding, why doesn't we support values beyond float limit - even for double values ?